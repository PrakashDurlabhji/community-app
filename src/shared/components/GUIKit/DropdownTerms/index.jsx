/* eslint-disable jsx-a11y/label-has-for */
/**
 * Dropdown terms component.
 */
import React, { useState, useRef, useEffect } from 'react';
import PT from 'prop-types';
import _ from 'lodash';
import { Creatable } from 'react-select';
import iconDown from 'assets/images/dropdown-arrow.png';
import './style.scss';

import { config } from 'topcoder-react-utils';

function DropdownTerms({
  terms,
  placeholder,
  label,
  required,
  onChange,
  errorMsg,
  addNewOptionPlaceholder,
}) {
  const [internalTerms, setInternalTerms] = useState(terms);
  const selectedOption = _.filter(internalTerms, { selected: true }).map(o => ({
    value: o.label,
    label: o.label,
  }));
  const [focused, setFocused] = useState(false);
  const delayedOnChange = useRef(
    _.debounce((q, cb) => cb(q), config.GUIKIT.DEBOUNCE_ON_CHANGE_TIME),
  ).current;
  const containerRef = useRef(null);
  let inputField;
  useEffect(() => {
    const selectInput = containerRef.current.getElementsByClassName('Select-input');
    const selectMenuOuter = containerRef.current.getElementsByClassName('Select-menu-outer');
    if (selectInput && selectInput.length) {
      const selectControl = containerRef.current.getElementsByClassName(
        'Select-control',
      );
      const height1 = selectMenuOuter && selectMenuOuter.length
        ? selectMenuOuter[0].offsetHeight
        : 0;
      const height2 = selectControl && selectControl.length
        ? selectControl[0].offsetHeight
        : 0;
      selectInput[0].style.top = focused ? `${height1 + height2 - 1}px` : '0';
      inputField = selectInput[0].getElementsByTagName('input');
      inputField[0].placeholder = focused ? addNewOptionPlaceholder : '';
      inputField[0].style.border = 'none';
      inputField[0].style.boxShadow = 'none';
    }
  }, [focused, selectedOption]);

  const CustomReactSelectRow = React.forwardRef(({
    className,
    option,
    children,
    onSelect,
  }, ref) => (children ? (
    <a
      ref={ref}
      role="button"
      className={className}
      onMouseDown={(event) => {
        event.preventDefault();
        event.stopPropagation();
        onSelect(option, event);
      }}
      title={option.title}
      tabIndex={-1}
    >
      {children}
    </a>
  ) : null));

  CustomReactSelectRow.defaultProps = {
    children: null,
    className: '',
    onSelect: () => {},
  };

  CustomReactSelectRow.propTypes = {
    children: PT.node,
    className: PT.string,
    onSelect: PT.func,
    option: PT.object.isRequired,
  };

  return (
    <div
      ref={containerRef}
      className="dropdownContainer"
      styleName={`container ${
        selectedOption && !!selectedOption.length ? 'haveValue' : ''
      } ${errorMsg ? 'haveError' : ''} ${_.every(internalTerms, { selected: true }) ? 'isEmptySelectList' : ''} ${focused ? 'isFocused' : ''}`}
    >
      <div styleName="relative">
        <Creatable
          onOpen={() => setFocused(true)}
          onClose={() => setFocused(false)}
          autosize={false}
          optionComponent={CustomReactSelectRow}
          options={internalTerms
            .map(o => ({ value: o.label, label: o.label }))
          }
          value={selectedOption}
          onInputKeyDown={(e) => {
            switch (e.keyCode) {
              case 13: // ENTER
                if (inputField && inputField && inputField[0]) {
                  const { value } = inputField[0];
                  if (
                    !value
                    || !value.trim()
                    || _.find(selectedOption, { label: value })
                    || _.find(internalTerms, { label: value })
                  ) {
                    e.preventDefault();
                  }
                }
                break;
              default:
            }
          }}
          onChange={(value) => {
            const newValues = _.filter(
              value,
              o => o.label && o.label.trim() && !_.find(internalTerms, { label: o.label }),
            ).map(o => ({ selected: true, label: o.label }));
            const newTerms = internalTerms
              .map(o => ({
                selected: !!_.find(value, { label: o.label }),
                label: o.label,
              }))
              .concat(newValues);
            setInternalTerms(newTerms);
            delayedOnChange(
              _.cloneDeep(newTerms),
              onChange,
            );
          }}
          placeholder={focused ? '' : `${placeholder}${placeholder && required ? ' *' : ''}`}
          clearable={false}
          backspaceRemoves={false}
          multi
          promptTextCreator={() => null}
          filterOptions={() => _.filter(
            internalTerms, t => !_.find(selectedOption, { label: t.label }),
          ).map(o => ({ value: o.label, label: o.label }))}
        />
        <img width="15" height="9" styleName="iconDropdown" src={iconDown} alt="dropdown-arrow-icon" />
      </div>
      {label ? (
        <span styleName="label">
          {label}
          {required ? <span>&nbsp;*</span> : null}
        </span>
      ) : null}
      {errorMsg ? <span styleName="errorMessage">{errorMsg}</span> : null}
    </div>
  );
}

DropdownTerms.defaultProps = {
  placeholder: '',
  label: '',
  required: false,
  onChange: () => {},
  errorMsg: '',
  addNewOptionPlaceholder: '',
};

DropdownTerms.propTypes = {
  terms: PT.arrayOf(
    PT.shape({
      label: PT.string.isRequired,
      selected: PT.bool.isRequired,
    }),
  ).isRequired,
  placeholder: PT.string,
  label: PT.string,
  required: PT.bool,
  onChange: PT.func,
  errorMsg: PT.string,
  addNewOptionPlaceholder: PT.string,
};

export default DropdownTerms;
