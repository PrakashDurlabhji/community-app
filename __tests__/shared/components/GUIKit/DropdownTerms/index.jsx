import React from 'react';
import Renderer from 'react-test-renderer/shallow';

import DropdownTerms from 'components/GUIKit/DropdownTerms';


const rnd = new Renderer();

it('Default render', () => {
  rnd.render((<DropdownTerms terms={[{ label: 'dropdownLabel', selected: true }]} />));
  expect(rnd.getRenderOutput()).toMatchSnapshot();
});
