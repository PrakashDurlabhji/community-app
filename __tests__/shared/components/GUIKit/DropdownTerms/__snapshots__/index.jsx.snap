// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default render 1`] = `
<div
  className="dropdownContainer src-shared-components-GUIKit-DropdownTerms-___style__container___oBZ2m src-shared-components-GUIKit-DropdownTerms-___style__haveValue___1Ppjh src-shared-components-GUIKit-DropdownTerms-___style__isEmptySelectList___CWafx"
>
  <div
    className="src-shared-components-GUIKit-DropdownTerms-___style__relative___EOspV"
  >
    <CreatableSelect
      autosize={false}
      backspaceRemoves={false}
      clearable={false}
      filterOptions={[Function]}
      isOptionUnique={[Function]}
      isValidNewOption={[Function]}
      menuRenderer={[Function]}
      multi={true}
      newOptionCreator={[Function]}
      onChange={[Function]}
      onClose={[Function]}
      onInputKeyDown={[Function]}
      onOpen={[Function]}
      optionComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "children": null,
            "className": "",
            "onSelect": [Function],
          },
          "propTypes": Object {
            "children": [Function],
            "className": [Function],
            "onSelect": [Function],
            "option": [Function],
          },
          "render": [Function],
        }
      }
      options={
        Array [
          Object {
            "label": "dropdownLabel",
            "value": "dropdownLabel",
          },
        ]
      }
      placeholder=""
      promptTextCreator={[Function]}
      shouldKeyDownEventCreateNewOption={[Function]}
      showNewOptionAtTop={true}
      value={
        Array [
          Object {
            "label": "dropdownLabel",
            "value": "dropdownLabel",
          },
        ]
      }
    />
    <img
      alt="dropdown-arrow-icon"
      className="src-shared-components-GUIKit-DropdownTerms-___style__iconDropdown___25Wtg"
      height="9"
      src="/community-app-assets/images/e740c51bb099110dff2900fa4fc94d72.png"
      width="15"
    />
  </div>
</div>
`;
